cloudinit/config/cc_kernel_modules.py

```diff
@@ -0,0 +1,399 @@
+    responsible for loading and enhancing kernel modules.
+
+    @param distro_cfg: Distro config
+
+    @raises RuntimeError when remove operation fails
+    """
+    CFG_FILES = distro_cfg["km_files"]
+
+    for (key, _) in sorted(CFG_FILES.items()):
+        file_path = CFG_FILES[key]["path"]
+        LOG.debug("Removing file %s", file_path)
+        try:
+            util.del_file(file_path)
+        except Exception as e:
+            raise RuntimeError(
+                f"Could not delete file {file_path}:{NL}{str(e)}"
+            ) from e
+
+
+def reload_modules(cloud: Cloud):
+    """Reload kernel modules
+
+    This function restarts service 'systemd-modules-load'
+    for reloading kernel modules
+
+    @param cloud: Cloud object
+
+    @raises RuntimeError
+    """
+    try:
+        cloud.distro.manage_service("restart", "systemd-modules-load")
+    except subp.ProcessExecutionError as e:
+        raise RuntimeError(
+            f"Could not restart service systemd-modules-load:{NL}{str(e)}"
+        ) from e
+
+
+def is_loaded(distro_cfg: dict, module_name: str):
+    """
+    Checks if a kernel module is already loaded
+
+    @param distro_cfg: Distro config
+    @param module_name: name of kernel module
+    """
+    loaded = False
+    try:
+        out = subp.subp(
+            distro_cfg["km_is_loaded_cmd"], capture=True, shell=True
+        )
+        if re.search("^" + module_name, out.stdout.strip()):
+            LOG.debug("Kernel module %s is loaded", module_name)
+            loaded = True
+    except subp.ProcessExecutionError as e:
+        util.logexc(
+            LOG,
+            f"Could not determine status of module {module_name}:{NL}{str(e)}",
+        )
+
+    return loaded
+
+
+def unload_modules(distro_cfg: dict, modules: list):
+    """Unloads a list of kernel module
+
+    This function unloads a list of kernel modules.
+
+    @param distro_cfg: Distro config
+    @param modules: list of module names
+
+    @raises RuntimeError
+    """
+    for module in set(modules):
+        cmd = distro_cfg["km_unload_cmd"].append(module)
+        try:
+            if is_loaded(distro_cfg, module):
+                LOG.debug("Unloading kernel module %s", module)
+                subp.subp(cmd)
+        except subp.ProcessExecutionError as e:
+            raise RuntimeError(
+                f"Could not unload kernel module {module}:{NL}{str(e)}"
+            ) from e
+
+
+def update_initial_ramdisk(distro_cfg: dict):
+    """Update initramfs for all installed kernels
+
+    @param distro_cfg: Distro config
+
+    @raises RuntimeError
+    """
+    try:
+        subp.subp(distro_cfg["km_update_cmd"])
+    except subp.ProcessExecutionError as e:
+        raise RuntimeError(
+            f"Failed to update initial ramdisk:{NL}{str(e)}"
+        ) from e
+
+
+def handle(
+    name: str, cfg: Config, cloud: Cloud, log: Logger, args: list
+) -> None:
+    kernel_modules_section = None
+
+    if "kernel_modules" in cfg:
```

Comment:
Having a look at this if-else block, if this module receives a config as:

```
cc_kernel_modules:
  kernel_modules:
```

Then the flow of execution is going to pass through the `if` block and `handle` will remove any existent configuration files but won't unload the modules, which sounds like an inconsistent state.

Unless I miss something, I think this case should also hit the else block and noop to avoid surprises. Something in the lines of:

```diff
diff --git a/cloudinit/config/cc_kernel_modules.py b/cloudinit/config/cc_kernel_modules.py
index 48499bbf9..01140ad1e 100644
--- a/cloudinit/config/cc_kernel_modules.py
+++ b/cloudinit/config/cc_kernel_modules.py
@@ -341,14 +341,11 @@ def update_initial_ramdisk(distro_cfg: dict):
 def handle(
     name: str, cfg: Config, cloud: Cloud, log: Logger, args: list
 ) -> None:
-    kernel_modules_section = None
-
-    if "kernel_modules" in cfg:
-        LOG.debug("Found kernel_modules section in config")
-        kernel_modules_section = util.get_cfg_option_list(
+    kernel_modules_section = util.get_cfg_option_list(
             cfg, "kernel_modules", []
         )
-    else:
+
+    if not kernel_modules_section:
         LOG.debug(
             "Skipping module named %s, no "
             "'kernel_modules' configuration found",
@@ -356,6 +353,8 @@ def handle(
         )
         return
 
+    LOG.debug("Found kernel_modules section in config")
+
     # check systemd usage
     if not uses_systemd:
         LOG.debug(
```

Reply:
Agree